cmd/internal/flowmode/cluster_builder.go

```diff
@@ -0,0 +1,179 @@
+package flowmode
+
+import (
+	"fmt"
+	stdlog "log"
+	"net"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/service/cluster"
+	"github.com/grafana/ckit/advertise"
+	"github.com/hashicorp/go-discover"
+	"github.com/hashicorp/go-discover/provider/k8s"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type clusterOptions struct {
+	Log     log.Logger
+	Metrics prometheus.Registerer
+	Tracer  trace.TracerProvider
+
+	EnableClustering bool
+	NodeName         string
+	AdvertiseAddress string
+	ListenAddress    string
+	JoinPeers        []string
+	DiscoverPeers    string
+	RejoinInterval   time.Duration
+}
+
+func buildClusterService(opts clusterOptions) (*cluster.Service, error) {
+	listenPort := findPort(opts.ListenAddress, 80)
+
+	config := cluster.Options{
+		Log:     opts.Log,
+		Metrics: opts.Metrics,
+		Tracer:  opts.Tracer,
+
+		EnableClustering: opts.EnableClustering,
+		NodeName:         opts.NodeName,
+		AdvertiseAddress: opts.AdvertiseAddress,
+		RejoinInterval:   opts.RejoinInterval,
+	}
+
+	if config.NodeName == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			return nil, fmt.Errorf("generating node name: %w", err)
+		}
+		config.NodeName = hostname
+	}
+
+	if config.AdvertiseAddress == "" {
+		// TODO(rfratto): allow advertise interfaces to be configurable.
+		addr, err := advertise.FirstAddress(advertise.DefaultInterfaces)
+		if err != nil {
+			return nil, fmt.Errorf("determining advertise address: %w", err)
+		}
+		config.AdvertiseAddress = fmt.Sprintf("%s:%d", addr.String(), listenPort)
+	} else {
+		config.AdvertiseAddress = appendDefaultPort(config.AdvertiseAddress, listenPort)
```

Comment:
I didn't see any place where we log the `AdvertiseAddress`? Seems like a useful thing to log.